packages/sign_in_with_apple/lib/sign_in_with_apple.dart

```diff
@@ -0,0 +1,124 @@
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
+
+export './sign_in_with_apple_button/sign_in_with_apple_button.dart';
+
+class SignInWithApple {
+  static const MethodChannel _channel =
+      const MethodChannel('de.aboutyou.mobile.app.sign_in_with_apple');
+
+  /// Request credentials from the system, preferring existing keychain credentials
+  /// over "Sign in with Apple"
+  ///
+  /// When no credentials are returned (e.g. also by the user cancelling), this is treated as
+  /// all other errors (just like the native API), and will throw an Exception
+  ///
+  /// Successful result will be either of type [AuthorizationAppleID] or [AuthorizationPassword]
+  static Future<Authorization> requestCredentials() async {
+    return _parseCredentialsResponse(
+      await _channel
+          .invokeMethod<Map<dynamic, dynamic>>('performAuthorizationRequest'),
+    );
+  }
+
+  static Future<CredentialState> getCredentialState(
+    String userIdentifier,
+  ) async {
+    return _parseCredentialState(
+      await _channel.invokeMethod<String>(
+        'getCredentialState',
+        <String, String>{'userIdentifier': userIdentifier},
+      ),
+    );
+  }
+}
+
+enum CredentialState {
+  authorized,
+  revoked,
+  notFound,
+}
+
+CredentialState _parseCredentialState(String credentialState) {
+  switch (credentialState) {
+    case 'authorized':
+      return CredentialState.authorized;
+
+    case 'revoked':
+      return CredentialState.revoked;
+
+    case 'notFound':
+      return CredentialState.notFound;
+
+    default:
+      throw Exception("Unsupported credential state $credentialState");
+  }
+}
+
+@immutable
+abstract class Authorization {}
+
+class AuthorizationAppleID implements Authorization {
+  AuthorizationAppleID({
+    @required this.userIdentifier,
+    @required this.givenName,
+    @required this.familyName,
+    @required this.email,
+  }) : assert(userIdentifier != null);
+
+  final String userIdentifier;
+
+  /// Can be `null`
+  final String givenName;
+
+  /// Can be `null`
+  final String familyName;
+
+  /// Can be `null`
+  final String email;
```

Comment:
In what circumstances could this be nil? In case the user is already registered?

Reply:
You only get e-mail and name once. If you would then show the sheet again with the app already being associated with the user's Apple ID, one only gets the user identifier back.

Reply:
Should we then maybe model this here just to make it clear?

Reply:
You mean 2 Apple ID credentials classes, with and without username/password/email?

Reply:
exactly, I think that would make it easier on our app side then to show the intent / possibilities

Reply:
Follow up created.